File: backend/src/test/java/com/learningportal/TestDataFactory.java
package com.learningportal;

import com.learningportal.entity.User;

import java.time.LocalDateTime;

/**
 * Factory class for creating test data.
 * Provides consistent test data across all tests.
 */
public class TestDataFactory {

    /**
     * Creates a test user with default values.
     */
    public static User createUser(String username, String email) {
        User user = new User();
        user.setUsername(username);
        user.setEmail(email);
        user.setPasswordHash("$2a$10$hashedpassword"); // BCrypt hash
        user.setCreatedAt(LocalDateTime.now());
        user.setUpdatedAt(LocalDateTime.now());
        return user;
    }

    /**
     * Creates a test user with custom password hash.
     */
    public static User createUser(String username, String email, String passwordHash) {
        User user = createUser(username, email);
        user.setPasswordHash(passwordHash);
        return user;
    }

    /**
     * Creates a test user with ID (for mocking persisted entities).
     */
    public static User createUserWithId(Long id, String username, String email) {
        User user = createUser(username, email);
        user.setId(id);
        return user;
    }
}

File: frontend/src/test/TestDataFactory.ts
/**
 * Factory for creating consistent test data across all tests.
 */

export interface TestUser {
  id: number;
  username: string;
  email: string;
  fullName?: string;
  createdAt: string;
  updatedAt: string;
}

export interface TestAuthResponse {
  token: string;
  user: TestUser;
}

export const TestDataFactory = {
  /**
   * Create a test user with default values.
   */
  createUser(
    username: string = 'testuser',
    email: string = 'test@example.com'
  ): TestUser {
    return {
      id: 1,
      username,
      email,
      fullName: 'Test User',
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString(),
    };
  },

  /**
   * Create a test user with custom ID.
   */
  createUserWithId(
    id: number,
    username: string = 'testuser',
    email: string = 'test@example.com'
  ): TestUser {
    return {
      ...this.createUser(username, email),
      id,
    };
  },

  /**
   * Create a test auth response.
   */
  createAuthResponse(user?: TestUser): TestAuthResponse {
    return {
      token: 'mock.jwt.token',
      user: user || this.createUser(),
    };
  },

  /**
   * Create a mock JWT token.
   */
  createJwtToken(username: string = 'testuser'): string {
    const header = btoa(JSON.stringify({ alg: 'HS256', typ: 'JWT' }));
    const payload = btoa(JSON.stringify({ sub: username, exp: Date.now() + 3600000 }));
    const signature = btoa('mock-signature');
    return `${header}.${payload}.${signature}`;
  },
};
